tasty-query/shared/src/main/scala/tastyquery/Symbols.scala

```diff
@@ -536,9 +470,49 @@ object Symbols {
     private[tastyquery] final def specialErasure(using Context): Option[() => ErasedTypeRef] =
       mySpecialErasure
 
-    protected[this] final def ensureDeclsInitialized()(using Context): Unit =
-      // ClassSymbols are always initialized when created
-      ()
+    // DeclaringSymbol implementation
+
+    private[Symbols] final def addDecl(decl: Symbol): Unit =
+      val set = myDeclarations.getOrElseUpdate(decl.name, new mutable.HashSet)
+      if decl.isType then assert(set.isEmpty, s"trying to add a second entry $decl for type name ${decl.name} in $this")
+      set += decl
+
+    private[Symbols] final def hasOverloads(name: SignedName): Boolean =
+      myDeclarations.get(name.underlying) match
+        case Some(decls) => decls.sizeIs > 1
+        case _           => false
+
+    final def getDecls(name: Name)(using Context): List[Symbol] =
+      name match
+        case name: SignedName => getDecl(name).toList
+        case _                => myDeclarations.get(name).fold(Nil)(_.toList)
+
+    final def getDecl(name: Name)(using Context): Option[Symbol] =
+      name match
+        case overloaded: SignedName =>
+          distinguishOverloaded(overloaded)
+        case name =>
+          myDeclarations.get(name) match
+            case Some(decls) =>
+              if decls.sizeIs == 1 then Some(decls.head)
+              else if decls.sizeIs > 1 then
+                throw MemberNotFoundException(this, name, s"unexpected overloads in $this: ${decls.mkString(", ")}")
+              else None
+            case _ => None
+
+    private final def distinguishOverloaded(overloaded: SignedName)(using Context): Option[Symbol] =
+      myDeclarations.get(overloaded.underlying) match
+        case Some(overloads) =>
+          overloads.find {
+            case decl: TermSymbol => decl.signature.exists(_ == overloaded.sig)
+            case _                => false
+          }
+        case None => None
+
+    final def declarations(using Context): List[Symbol] =
+      myDeclarations.values.toList.flatten
```

Comment:
seeing `ensureDeclsInitialized` dropped (even tho it was a no-op) made me think that we don't have a check to ensure these methods are not called during initialisation

Reply:
I can try to come up with something here if you want. We could have a flag that is set when doing `checkCompleted()`. Before it's set, we cannot call `getDecl`/`declarations`. After it's set, we cannot call `addDecl`. Would that make sense?